chore: [FC-0074] modernize repo and readme with latest updates (#7)
eduNEXT/openedx-events-2-zapier: README.rst

Purpose
********

This repository demonstrates how to connect Open edX registration, enrollment, and grade change events to external tools via Zapier, enabling easier automation workflows through this third-party service.

Open edX Events are a powerful feature that allows developers to listen to key events in the Open edX platform and trigger custom actions based on them. This can be useful for a variety of use cases, such as:

- Sending welcome emails to new users
- Logging new enrollments to external CRMs
- Triggering events like email follow-ups for grade updates

By sending key event data to Zapier, Open edX users can leverage the integration ecosystem of Zapier without additional development effort.

Getting Started with Development
********************************

Please see the Open edX documentation for `guidance on Python development`_ in this repo.

.. _guidance on Python development: https://docs.openedx.org/en/latest/developers/how-tos/get-ready-for-python-dev.html

Deploying
*********

> See the Usage section below for instructions on how to deploy this plugin. Also, see the `Tutor documentation`_ for more information on deploying extra requirements.

Reviewer: Missing link to **Tutor documentation**

Django Signal Handlers
~~~~~~~~~~~~~~~~~~~~~

In the file `handlers.py`_, handlers listen to Django signals using the standard `receiver`_ decorator:

.. code-block:: python

    from django.dispatch import receiver
    from openedx_events.signals import STUDENT_REGISTRATION_COMPLETED

    @receiver(STUDENT_REGISTRATION_COMPLETED)
    def send_user_data_to_webhook(signal, sender, user, metadata, **kwargs):
        zapier_payload = {
            "user": asdict(user),
            "event_metadata": asdict(metadata),
        }
        requests.post(
            settings.ZAPIER_REGISTRATION_WEBHOOK,
            flatten_dict(zapier_payload),
            timeout=ZAPIER_REQUEST_TIMEOUT,
        )

- The ``receiver`` decorator listens to the ``STUDENT_REGISTRATION_COMPLETED`` signal.
- The handler function ``send_user_data_to_webhook`` extracts the user and metadata from the signal.
- The ``ZAPIER_REGISTRATION_WEBHOOK`` URL is configured as a Django settings by using a `Tutor plugin`_.
- The extracted data is formatted into a payload and sent to the Zapier webhook for further processing.

> App Configuration (`apps.py`)

Reviewer: ```suggestion
App Configuration (``apps.py``)
```

> App Configuration (`apps.py`)

Reviewer: ```suggestion
App Configuration (``apps.py``)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The Django app is configured using an `AppConfig` to automatically register handlers on startup.

Reviewer: ```suggestion
The Django app is configured using an ``AppConfig`` to automatically register handlers on startup.
```

.. code-block:: python

    class OpenedxEvents2ZapierConfig(AppConfig):
        name = "openedx_events_2_zapier"

        def ready(self):
            from openedx_events_2_zapier import handlers

Usage
-----

To use this plugin, follow these steps:

> 1. Install the plugin in your Open edX image using Tutor's `OPENEDX_EXTRA_PIP_REQUIREMENTS` configuration setting:

Reviewer: ```suggestion
1. Install the plugin in your Open edX image using Tutor's ``OPENEDX_EXTRA_PIP_REQUIREMENTS`` configuration setting:
```

> .. code-block:: yaml

Reviewer: Please indent the code-block

    OPENEDX_EXTRA_PIP_REQUIREMENTS:
    - git+https://github.com/edunext/openedx-events-2-zapier.git@X.Y.Z

2. Launch the Open edX platform to apply the changes:

> .. code-block:: bash

Reviewer: Please indent the code-block

How to Extend this Repository
-----------------------------

This repository is a starting point for Open edX developers:

- You can add new event handlers by following the structure in `handlers.py`_.
- Custom logic can be implemented to fit your organization's data flow requirements using Zapier, third-party APIs, or internal services.

For details on extending Open edX with Open edX Events, see also:

- `Open edX Events Documentation`_
- `Hooks Extension Framework`_

The openedx-events-2-zapier repository is here to make integrations simple and sustainable, giving developers the tools to create effective Open edX workflows with external services like Zapier.

More Help
=========

If you're having trouble, we have discussion forums at
https://discuss.openedx.org where you can connect with others in the
community.

Our real-time conversations are on Slack. You can request a `Slack
invitation`_, then join our `community Slack workspace`_.

For anything non-trivial, the best path is to open an issue in this
repository with as many details about the issue you are facing as you
can provide.

> https://github.com/openedx/openedx-events-samples/issues

Reviewer 2: This isn't a real link, did you mean to point to this repo?

For more information about these options, see the `Getting Help <https://openedx.org/getting-help>`__ page.

.. _Slack invitation: https://openedx.org/slack
.. _community Slack workspace: https://openedx.slack.com/

License
*******

The code in this repository is licensed under the AGPL 3.0 unless
otherwise noted.

Please see `LICENSE.txt <LICENSE.txt>`_ for details.

Contributing
************

Contributions are very welcome.
Please read `How To Contribute <https://openedx.org/r/how-to-contribute>`_ for details.

This project is currently accepting all types of contributions, bug fixes,
security fixes, maintenance work, or new features.  However, please make sure
to discuss your new feature idea with the maintainers before beginning development
to maximize the chances of your change being accepted.
You can start a conversation by creating a new issue on this repo summarizing
your idea.

The Open edX Code of Conduct
****************************

All community members are expected to follow the `Open edX Code of Conduct`_.

.. _Open edX Code of Conduct: https://openedx.org/code-of-conduct/

People
******

> This repository is currently being maintained by the eduNEXT team. See the CODEOWNERS file for details.

Reviewer: ```suggestion
This repository is currently being maintained by the edunext team. See the `CODEOWNERS <.github/CODEOWNERS>`_ file for details.
```

        ZAPIER_GRADE_WEBHOOK = "https://hooks.zapier.com/hooks/catch/<account>/<webhook>/"
        """
            )
        )

 .. code-block:: bash

      tutor plugins enable zapier

4. Configure Zapier webhooks to receive the event data, follow the instructions available in the Zapier documentation.
5. Trigger the events by registering a new user, enrolling in a course, or updating a grade in the Open edX platform.

To send event data to other services or APIs, simply configure more webhooks in the Django settings. The handlers are intentionally generic, ensuring they work seamlessly with different kinds of services. You can also add more event handlers to the `handlers.py`_ file to listen to additional events.

How to Extend this Repository
-----------------------------

This repository is a starting point for Open edX developers:

- You can add new event handlers by following the structure in `handlers.py`_.
- Custom logic can be implemented to fit your organization's data flow requirements using Zapier, third-party APIs, or internal services.

For details on extending Open edX with Open edX Events, see:

- `Open edX Events Documentation`_
- `Hooks Extension Framework`_

> The openedx-events-2-zapier repository is here to make integrations simple and sustainable, giving developers the tools to create effective Open edX workflows. Use Open edX Events and Zapier to extend the platform's capabilities and unlock new automation possibilities!

Author: ```suggestion
The openedx-events-2-zapier repository is here to make integrations simple and sustainable, giving developers the tools to create effective Open edX workflows with external services like Zapier.
```

        }
        requests.post(
            settings.ZAPIER_REGISTRATION_WEBHOOK,
            flatten_dict(zapier_payload),
            timeout=ZAPIER_REQUEST_TIMEOUT,
        )

- The ``receiver`` decorator listens to the ``STUDENT_REGISTRATION_COMPLETED`` signal.
- The handler function ``send_user_data_to_webhook`` extracts the user and metadata from the signal.
- The ``ZAPIER_REGISTRATION_WEBHOOK`` URL is configured as a Django settings by using a `Tutor plugin`_.
- The extracted data is formatted into a payload and sent to the Zapier webhook for further processing.

App Configuration (`apps.py`)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The Django app is configured using an `AppConfig` to automatically register handlers on startup.

.. code-block:: python

    class OpenedxEvents2ZapierConfig(AppConfig):
        name = "openedx_events_2_zapier"

        def ready(self):
            from openedx_events_2_zapier import handlers

Usage
-----

To use this plugin, follow these steps:

1. Install the plugin in your Open edX image using Tutor's `OPENEDX_EXTRA_PIP_REQUIREMENTS` configuration setting:

.. code-block:: yaml

    OPENEDX_EXTRA_PIP_REQUIREMENTS:

>     - git+https://github.com/edunext/openedx-events-2-zapier.git@main

Author: ```suggestion
    - git+https://github.com/edunext/openedx-events-2-zapier.git@X.Y.Z
```
